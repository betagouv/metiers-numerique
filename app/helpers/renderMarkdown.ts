import { handleError } from '@common/helpers/handleError'
import React from 'react'
import rehypeReact from 'rehype-react'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { unified } from 'unified'

import type { ReactElement, JSXElementConstructor } from 'react'

export function renderMarkdown(
  markdownSource: string,
): ReactElement<unknown, string | JSXElementConstructor<any>> | null {
  try {
    const content = unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeReact, { createElement: React.createElement })
      .processSync(markdownSource).result

    return content
  } catch (err) /* istanbul ignore next */ {
    handleError(err, 'app/helpers/renderMarkdown()')

    return null
  }
}
