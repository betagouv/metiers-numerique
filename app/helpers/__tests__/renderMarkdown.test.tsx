/**
 * @jest-environment jsdom
 */

import { renderMarkdown } from '../renderMarkdown'

describe.only('app/helpers/renderMarkdown()', () => {
  test(`with a paragraph`, () => {
    const markdownSource = [
      'A paragraph.',
      '',
      'A **bold** and _italic_ text with a [link](https://example.org).',
      '',
      '- A',
      '- bullet,',
      '- list.',
      '',
      '1. An',
      '2. ordered,',
      '3. list.',
    ].join('\n')

    const result = renderMarkdown(markdownSource)

    expect(result).toMatchSnapshot()
  })
})
