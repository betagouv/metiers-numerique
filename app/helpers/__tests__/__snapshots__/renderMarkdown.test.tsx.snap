// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/helpers/renderMarkdown() with a paragraph 1`] = `
<div>
  <p>
    A paragraph.
  </p>
  

  <p>
    A 
    <strong>
      bold
    </strong>
     and 
    <em>
      italic
    </em>
     text with a 
    <a
      href="https://example.org"
    >
      link
    </a>
    .
  </p>
  

  <ul>
    

    <li>
      A
    </li>
    

    <li>
      bullet,
    </li>
    

    <li>
      list.
    </li>
    

  </ul>
  

  <ol>
    

    <li>
      An
    </li>
    

    <li>
      ordered,
    </li>
    

    <li>
      list.
    </li>
    

  </ol>
</div>
`;
